### PR TITLE
Update documentation for AsyncRead and AsyncWrite

### PR DESCRIPTION
**Pull request by zstewar1**
## Motivation

Documentation of AsyncRead and AsyncWrite do not reflect current API, referencing `Err(e) if e.kind() == ErrorKind::WouldBlock` as a way of identifying incomplete async work. The current implementation uses the `Async` enum in the `Ok` variant of `Result` rather than a special Error type to identify incomplete async work.

## Solution

Updated documentation to reflect the current API.

### CODE REVIEW

File: tokio-io/src/async_read.rs [hunk 1]
```diff
@@ -16,15 +16,16 @@ use split::{ReadHalf, WriteHalf};
...
+///   and placed into the output buffer, where `n` == 0 implies that EOF has
+///   been reached.
 ///
-/// * `Err(e) if e.kind() == ErrorKind::WouldBlock` means that no data was read
-///   into the buffer provided. The I/O object is not currently readable but may
-///   become readable in the future. Most importantly, **the current future's
-///   task is scheduled to get unparked when the object is readable**. This
-///   means that like `Future::poll` you'll receive a notification when the I/O
-///   object is readable again.
+/// * `Ok(Async::WouldBlock)` means that no data was read into the buffer
```
Comment (tobz) [inline tokio-io/src/async_read.rs]: This is incorrect.  It should be `Async::NotReady`.

File: tokio-io/src/async_write.rs [hunk 2]
```diff
@@ -13,14 +13,15 @@ use AsyncRead;
...
+/// * `Ok(Async::Ready(n))` means that `n` bytes of data was immediately
+///   written.
 ///
-/// * `Err(e) if e.kind() == ErrorKind::WouldBlock` means that no data was
-///   written from the buffer provided. The I/O object is not currently
-///   writable but may become writable in the future. Most importantly, **the
-///   current future's task is scheduled to get unparked when the object is
-///   readable**. This means that like `Future::poll` you'll receive a
-///   notification when the I/O object is writable again.
+/// * `Ok(Async::WouldBlock)` means that no data was written from the buffer
```
Comment (tobz) [inline tokio-io/src/async_write.rs]: This is incorrect.  It should be `Async::NotReady`.

**Comment by carllerche** [comment]
Yes, @tobz is correct

**Comment by zstewar1** [comment]
Actually, I think I misread. This is referring to the underlying `std::io::Write.write()` function call, which does return `std::io::Error` with `kind() == std::io::ErrorKind::WouldBlock`.

So this documentation is technically correct, but it still seems... confusing to someone just coming on it. It could use some clarification that it's referring to what `std::io::Write::write()` returns, and not what is returned from the extra methods added to `AsyncWrite`.

**Comment by tobz** [comment]
So, I understand where you're coming from.

The documentation is for `Async[Read|Write]` but does talk about the underlying blocking API surfaces.  It's slightly confusing.  I do still believe, however, that your change is correct, insofar as it matches the paragraph above and below it i.e. referring to the `Async` enum variants correctly.  Any asynchronous read/write on these traits _will_ return the types that are mentioned, and so they aren't one-to-one with `read`/`write`.

The part where I agree with you is that the documentation _as is_ mentions `read` and `write`, which, while it is calling those under the hood, actual users aren't calling those methods directly, at least not if they're interoperating with futures and combinators and what have you.  This blurs the line as to what is happening both on the surface and under the hood.

Maybe we could slightly reword those sections to talk about the behavior of _reads_ on an `AsyncRead` object, or _writes_ on an `AsyncWrite` object, to talk about the intended behavior exposed by the traits instead of trying to map precisely to how they differ from their blocking versions in the standard API.

Thoughts?

**Pull request merged**